### ISSUE TITLE
[Good First Issue]: Operation list to be supported

### ISSUE THREAD

**Issue by mvafin** [openvinotoolkit/openvino#28584]
Hello, everyone. I noticed an interest in PyTorch GFI. Right now we have the following list of operations that can be enabled:

> **If you want to enable one or more of these operations, assign yourself to the sub-issue.**

Please note that not all operations can be easily enabled and the task may become more exploratory than a simple implementation.

Example PRs:

- https://github.com/openvinotoolkit/openvino/pull/28933

<details>
  <summary>Do not work on these for now</summary>

  ```
aten::dict
aten::pad_sequence
aten::to_dense
aten::to_padded_tensor
aten::to_mkldnn
prim::ConstantMKLDNNTensor
prim::mkldnn_convolution
torch_ipex::ipex_linear
  ```
</details>

**Comment by itsbharatj** [comment]
.take

**Comment by itsbharatj** [comment]
Hey @mvafin thank you for this list for all the operations that have to be enabled. I would be very excited to work and implement many of these components. 

Just for my understanding, is there any avaibilbe resource for the formula or description of the respective items?

**Comment by itsbharatj** [comment]
@darksapien23151 Definately! Would love to

**Comment by mvafin** [comment]
@itsbharatj @darksapien23151 Please take one or two operations to work on. Do not take this issue. New issue will be created for operations you took.
I would recommend investigating the operation using pytorch documentation and code.
Please also write short description of how would you support this operation, write a formula you would use, OpenVINO operations that will be used in decomposition, etc.

**Comment by geeky33** [comment]
@p-wysocki @mvafin 

Greetings to both of you sir! 
I am eager to work for this issue
since its a long enough 
@p-wysocki sir asked me that I can work out on this issue's some of the selected operators

May I get the chance to work on 

aten::equal
aten::Delete
aten::dict
aten::dot
aten::stack
aten::str
aten::take


Waiting for the positive response from both of you! 

Thank you and Regards!!! :)

**Comment by mvafin** [comment]
I removed `aten::equal` as it was already added.
`aten::dict` requires support for dicts in the frontend, so I wouldn't recommend starting from it. I would recommend you start from `aten::dot` as it seems to be easier operation

**Comment by geeky33** [comment]
okay sir
should I raise the issue and create the PR for it as it might get convenient for you to check on what selected operators I worked on!

Thanks for assigning me the issue!!

Regards!!

**Comment by gatbontonp** [comment]
Hi @mvafin,

I picked up a few operations:
aten::stack
aten::vstack
aten::hstack
I also started writing aten::nan_to_num which would be combining the openvino operations IsNaN, IsInf, and Selects, but I see that @darksapien23151 also requested it.

**Comment by code-dev05** [comment]
Hi @mvafin 
I would like to take on the `aten::logit` function for now.
It should be done using the log and subtract function.
Should I make a new issue for this function or just reference this issue in the pr?

**Comment by rtruong3775** [comment]
Hi @mvafin are there any functions left I can help contribute towards?

**Comment by Mohamed-Ashraf273** [comment]
Hey @mvafin 
I hope you're doing great! I've picked out some simple operations to get started with:
aten::_to_copy
aten::__lshift__ 
aten::__rshift__

Can I contribute to this issue?

**Comment by carlozamu** [comment]
.take @mvafin 
 I would like to work on `aten::rad2deg`. 
My plan is to implement it by applying a simple element-wise multiplication by the constant 
 180
----
  𝜋

to convert radians to degrees, using the corresponding OpenVINO operations for multiplication and constant creation.
Tell me if is ok to start!!

**Comment by mvafin** [comment]
> .take [@mvafin](https://github.com/mvafin) I would like to work on `aten::rad2deg`. My plan is to implement it by applying a simple element-wise multiplication by the constant 180 ---- 𝜋 to convert radians to degrees, using the corresponding OpenVINO operations for multiplication and constant creation. Tell me if is ok to start!!

Sounds, good. You can start, here is the issue for you #28873

**Comment by mvafin** [comment]
> Hey [@mvafin](https://github.com/mvafin) I hope you're doing great! I've picked out some simple operations to get started with: aten::**lshift** aten::**rshift**
> 
> Can I contribute to this issue?

Yes, here is the issue for you: #28687 
This is a bitwise shift operations, so please also do `aten::bitwise_right_shift` and `aten::bitwise_left_shift`

**Comment by mvafin** [comment]
> Hi [@mvafin](https://github.com/mvafin),
> 
> I picked up a few operations: aten::stack aten::vstack aten::hstack I also started writing aten::nan_to_num which would be combining the openvino operations IsNaN, IsInf, and Selects, but I see that [@darksapien23151](https://github.com/darksapien23151) also requested it.

Created an issue for you #28875. `aten::stack` is actually already supported, so I deleted it. The hstack and vstack is a versions of it.

**Comment by mvafin** [comment]
@geeky33 @darksapien23151 Are you working on any operation already, do you want me to create an issue for you?

**Comment by mvafin** [comment]
@carlozamu @Mohamed-Ashraf273 @gatbontonp please `.take` your issues

**Comment by cyberholics** [comment]
Hello sir @mvafin I would like to enable aten::ravel, which flattens a tensor to a 1D array. I plan to implement this using OpenVINO’s Reshape operation. Specifically, I will take the input tensor and apply Reshape with a shape of [-1] to achieve a flattened output while preserving the original data layout.

OpenVINO operations used:

ov::op::v1::Reshape (to reshape the tensor to a 1D array).
ov::op::v0::Constant (to define the new shape [-1]).

**Comment by mvafin** [comment]
> Hello sir [@mvafin](https://github.com/mvafin) I would like to enable aten::ravel, which flattens a tensor to a 1D array. I plan to implement this using OpenVINO’s Reshape operation. Specifically, I will take the input tensor and apply Reshape with a shape of [-1] to achieve a flattened output while preserving the original data layout.
> 
> OpenVINO operations used:
> 
> ov::op::v1::Reshape (to reshape the tensor to a 1D array). ov::op::v0::Constant (to define the new shape [-1]).

@cyberholics here is an issue for you #28902

**Comment by geeky33** [comment]
@mvafin 
Greetings,
I have created the issue and assigned it to myself

could you please check the issue created by me :
#28707

if it doesn't seem applicable
you can create an issue and I will assign it to myself

Thanks.

**Comment by Mohamed-Ashraf273** [comment]
hi @mvafin ,
I finished sub-issue #28874.
I made a PR and it's ready for review #28939

**Comment by janviisonii23** [comment]
.take
Hello @mvafin,I would like to work on aten::binary_cross_entropy_with_logits, which computes binary cross-entropy loss directly from raw logits. First, I will apply a sigmoid activation to transform the logits into probability values. Then, I will compute the binary cross-entropy loss using logarithmic operations, element-wise multiplication, and reduction.

Operations used:
ov::op::v0::Sigmoid → Converts logits into probabilities.
For the BCE loss the operations used would be 
ov::op::v0::Log → Computes the logarithm of probabilities.
ov::op::v1::Subtract → Calculates (1 - p) and (1 - y).
ov::op::v1::Multiply → Performs element-wise multiplication.
ov::op::v0::Negative → Negates the sum of the loss terms.
ov::op::v1::ReduceMean → Computes the final loss by averaging across elements.

**Comment by itsbharatj** [comment]
@mvafin 
I would like to work on the following operations: 

```
aten::replace
aten::reverse
```

Can you please raise an issue for me to start working on them.

**Comment by wang-zhuoran** [comment]
Hi, I would like to work on `aten::polar`. @mvafin Can you assign these two to me? Thanks!

**Comment by Jyc323** [comment]
Hello, @mvafin
I would like to work on the following operations:

**aten::cross_entropy_loss**

Taking logits, targets and reduction method as the input, I will use `ov::opset10::Softmax` and `ov::opset10::Log` to calculate the log probability, and use `ov::opset10::Negative` to calculate negative loglikelihood. Finally use `ov::opset10::ReduceMean` for reduction.

**aten::smooth_l1_loss**

Taking predicted values, targets, beta, I will use `ov::opset10::Subtract` and `ov::opset10::Abs` to compute absolute error, then compute quadratic term and linear term. Finally select between different terms, and do the reduction.

Could you please create an issue and assign to me? Thanks!

**Comment by A-l-Mahi** [comment]
Hi @mvafin,  

I noticed that `aten::copy_()`, which performs in-place tensor copying in PyTorch, is not in the provided list. In contrast, `aten::_to_copy` creates a new tensor on a specific device.  

I’d love to work on adding `aten::copy_()`. If you approve, I’ll share my implementation plan shortly.  

Thanks!

**Comment by vijaykr338** [comment]
hi @mvafin @mlukasze 

I have worked on aten::str, aten::take and aten::Delete 
but I am having issues implementing tests for str and Delete - #29479  PR

I am waiting for a review

so in the meantime, can I work on aten::angle and aten::randperm? 

aten::angle
the implementation as we just have to extract the real and imaginary components in a floating-point type and then use the arctangent function 

aten::randperm
we simply have to generate numbers from 0 to n-1 

can you please raise an issue, so I can start working on it

**Comment by Utsuk7** [comment]
Hello @mvafin ,

I would like to work on the following issues:

1) aten::reverse
2) quantized::linear_relu

I plan to implement a custom reverse_tensor function for reversing tensors.
Do you have any thoughts? I’m open to suggestions.

Could you please assign these to me?

**Comment by hibahkhan2022** [comment]
@mvafin I would like to take

`aten::tensordot
aten::to_dense`

I've checked the repo, and don't see any requests for it. Is it safe to implement?

**Comment by johnrhimawan** [comment]
Hi @mvafin ~ I'm interested in taking `aten::smooth_l1_loss`, it seems that there's no requests for this yet. Would it be okay for me to implement?

My plan is to refer to the [PyTorch SmoothL1Loss](https://pytorch.org/docs/stable/generated/torch.nn.SmoothL1Loss.html) and implement the features here. So the code would be something like
1. Take predictions and grouth truth as input
2. Supports the beta parameter for smoothing
3. Supporting the different reduction modes (e.g. `none`, `mean`, `sum`)

The operations I would be using are:
* `ov::op::v0::Abs>`
* `ov::op::v0::Subtract>`
* `ov::op::v1::Less>`
* `ov::op::v1::Multiply>`
* `ov::op::v1::ReduceSum>`
* etc

to implement the functionality

**Comment by mvafin** [comment]
@johnrhimawan @hibahkhan2022 @Utsuk7  @A-l-Mahi @vijaykr338 @Jyc323 @wang-zhuoran  Please create issues by your own as there is no way to assign an issue to you on creation. I will link those issues here.

**Comment by vijaykr338** [comment]
@mvafin I have opened 
#29479  - aten::str, aten::take and aten::delete
#29585  - aten::randperm and aten:: polar

I will give an update for the first PR very soon and the second one is in waiting for a review

please link this issue here
#29547 - aten::randperm and aten::polar 


meanwhile I would like to work on 
quantised::relu6 and prim::dtype 

#29649 - prim::dtype and quantised::relu6
(please link this as well)

I have already created an issue and assigned it to myself 

they are both very simple functions 

prim::dtype - returns a constant tensor that represents the data type of the input tensor.
quantised::relu6 - applies a ReLU6 activation function that clamps the input tensor values between 0 and 6.

additionally, I have been taught about ReLU in the university, so I am interested in working on it

**Comment by A-l-Mahi** [comment]
> [@mvafin](https://github.com/mvafin) I would like to take
> 
> `aten::tensordot aten::to_dense`
> 
> I've checked the repo, and don't see any requests for it. Is it safe to implement?

Hey, kindly check my comment 2 days before yours, as I have already requested to be assigned the task for aten::tensordot. 
Thank you.

**Comment by sallycaoyu** [comment]
Hi @mvafin I opened an issue #29658 for supporting `aten::clear` and assigned to myself.

**Comment by BIG-FAT-PENGUIN** [comment]
Hi @mvafin,

I’d like to contribute to this issue by working on **aten::unique_consecutive** and **aten::truncate.**

### **Plan:**

  - Analyze how **aten::unique_consecutive** and **aten::truncate** are implemented in **PyTorch**.
  - Investigate how these operations can be adapted or if similar functionality exists in **OpenVINO,** ensuring 
    compatibility with the backend.
  - Implement and run test cases to validate correctness and performance.

From my findings, these operations haven't been assigned yet. If I missed anything or if there are any constraints, please let me know. Looking forward to your feedback!

**Comment by hibahkhan2022** [comment]
@mvafin Since the previous requested have already been taken, I'd like to take:

aten::find
aten::format

**Comment by Javier-lz** [comment]
.take 

Could I do aten::pop? 

I am not completely sure how complex it is so any recommendation would be appreciated.

**Comment by Samay10** [comment]
.take 

I would like to take up aten::tensordot.

## Implementation Plan  
- **Understanding PyTorch Behavior**  
  - `torch.tensordot(A, B, dims)` computes a dot product along selected axes.  
  - It sums over specified dimensions after aligning the input tensors.  

- **Mapping to OpenVINO Operations**  
  - We can decompose `tensordot` using the following OpenVINO operations:  
    - **Transpose**: Align tensors to match the expected reduction dimensions.  
    - **MatMul**: Perform matrix multiplication between transformed tensors.  
    - **ReduceSum**: Sum over the specified dimensions to finalize the result.  

- **Frontend Integration**  
  - Add `aten::tensordot` support in OpenVINO's PyTorch frontend.  
  - Implement necessary shape transformations for alignment.  

- **Optimization & Validation**  
  - Optimize memory usage and execution efficiency.  
  - Write unit tests to verify correctness against PyTorch outputs.

**Comment by Javier-lz** [comment]
Hello I'm trying to find information on aten::pop to contribute to this issue, however I can not find any information of the function here: [https://pytorch.org/docs/stable/torch.compiler_ir.html#core-aten-ir](https://pytorch.org/docs/stable/torch.compiler_ir.html#core-aten-ir), where I believe it should be, since some of the already implemented functions are in the list. 

Does anyone know where else to find information?
I also searched in the PyTorch repository in here, [PyTorch repository](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/README.md). 

Thank you

**Comment by mvafin** [comment]
> Hello I'm trying to find information on aten::pop to contribute to this issue, however I can not find any information of the function here: https://pytorch.org/docs/stable/torch.compiler_ir.html#core-aten-ir, where I believe it should be, since some of the already implemented functions are in the list.
> 
> Does anyone know where else to find information? I also searched in the PyTorch repository in here, [PyTorch repository](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/README.md).
> 
> Thank you

This is Core Aten IR, where the list of operations is reduced, as other operations can decompose to them. However the decomposition is implemented for Graph FX representation, not TorchScript.
I suspect that aten::pop is the list.pop function and it may be tricky to implement it since  openvino doesn't support lists natively, so I would recommend looking for other operations.

**Comment by Samay10** [comment]
Greetings @mvafin!

Apologies for the previous takes. Both turned out to be assigned.
I would like to work on quantized::linear_relu as this has not been assigned yet.

Technique to be used:
Implementing quantized::linear_relu by fusing two operations: a quantized linear transformation (MatMul + Add) and a ReLU activation. This is done using OpenVINO’s MatMul, Add, and Relu operations to efficiently map the PyTorch function while maintaining computational efficiency.

If it looks good, I can go ahead and take it up.

**Comment by Samay10** [comment]
.take quantized::linear_relu

**Comment by p-wysocki** [comment]
### **IMPORTANT**

Hello everyone,

We decided to change the method of managing the operators subtask list, since this issue grew into chaos. From now on all the tasks are already created and are to be treated as regular Good First Issue tasks. **The current Issue is now only for organization**, please do not assign yourselves to it, do not add comments about picking tasks (general comments/questions about operators are of course welcome).

### Please do not assign yourselves to this issue, do not create new issues, just choose a sub-issue you're interested in and pick it.

**Comment by Javier-lz** [comment]
Hello, where can I find the documentation for the Pytorch functions? I have found [IR's ](https://pytorch.org/docs/stable/torch.compiler_ir.html#core-aten-ir) and the Pytorch repository [Pytorch](https://github.com/pytorch/pytorch/tree/main/aten/src/ATen), but most of the unassigned functions are either not in here, are not beginner friendly or the documentation doesn't seem to be right. 

Thank you

**Comment by saadkhi** [comment]
@mvafin this issue is assigned to me, can you brief me on which file I have to look upon to start the work.

**Comment by EdOgbalor** [comment]
Hi. Please does anyone know where I can find a description for the problem. [[Good First Issue]: Support torch_scatter::segment_mean_csr #29734](https://github.com/openvinotoolkit/openvino/issues/29734)

**Comment by Shruti2301** [comment]
Hi @mvafin and team,

I’m interested in contributing to this project as part of my preparation for Google Summer of Code (GSoC) 2025, and I’d like to start with some of the more beginner-friendly operations to get familiar with the codebase and workflow.

After going through the current list of available tasks, I’m particularly interested in working on the following:
	1.	aten::pop (#29710) – Appears to mirror Python’s list.pop() method. Likely minimal tensor logic and mostly frontend integration.
	2.	aten::clear (#29658) – Similar to list.clear() in Python. Should be simple to trace using a scripted model.
	3.	aten::items (#29701) – Could relate to the dictionary .items() method. Might involve working with Python native types, not tensors.
	4.	prim::isinstance (#29723) – A mapping for Python’s built-in isinstance(). Doesn’t seem to involve tensor operations.
	5.	prim::data (#29721) – Based on previous comments, this may map to an Identity op, making it a straightforward implementation.

These look like a great starting point to understand the IR tracing and operator support mechanism in PyTorch frontend for OpenVINO.

Please let me know if it’s okay to begin exploring and drafting implementations for these. I’d also be happy to share formulas, test models, and expected mappings once I start working on them.

Thanks in advance! Looking forward to contributing for GSOC.

Regards,
Shruti Mandaokar

**Comment by Shruti2301** [comment]
.take

**Comment by Shruti2301** [comment]
Support Plan for Selected Operations

1. aten::pop (Equivalent to list.pop())
	•	Behavior: Removes and returns an element from a list.
	•	Implementation: Use Gather to extract the element and Squeeze to remove it.

2. aten::clear (Equivalent to list.clear())
	•	Behavior: Empties the list.
	•	Implementation: Return an empty list.

3. aten::items (Equivalent to dict.items())
	•	Behavior: Returns key-value pairs.
	•	Implementation: Use Split to separate keys and values.

4. prim::isinstance (Equivalent to isinstance(obj, type))
	•	Behavior: Checks if an object is of a given type.
	•	Implementation: Use Equal + logical operations.

5. prim::data (Implemented via Identity)
	•	Behavior: Returns input unchanged.
	•	Implementation: Use Identity operation.
